torchinductor/lowering.py

```diff
@@ -469,6 +470,62 @@ def slice_(x, dim, start, end, step=1):
     return TensorBox(ir.SliceView.create(x.data, dim, start, end, step))
 
 
+@register_lowering(aten.roll)
+def roll(a, shifts, dims=tuple()):
+    """
+    This is based on torch._refs.roll(), but uses ir.ModularIndexing().
+
+    We can't use the ref here because it is based on multiple calls to
+    torch.cat() that this will result in terrible code.
+    """
+    # ATen specifies int[1] type for shifts and dims which expands integers to tuples of length 1
+    if not isinstance(shifts, Iterable):
+        shifts = (shifts,)
+    if not isinstance(dims, Iterable):
+        dims = (dims,)
+    dims = [_validate_dim(a, d) for d in dims]
+
+    len_shifts = len(shifts)
+    len_dims = len(dims)
+    if len_shifts != 1 or len_dims != 1:
+        if len_shifts == 0:
+            raise RuntimeError("`shifts` required")
+        # Takes care of the case when dims is not specified (default)
+        # By default, the tensor is flattened before shifting, after which the original shape is restored
+        if len_dims == 0 and len_shifts == 1:
+            flat = view(a, [sympy_product(a.get_size())])
+            rolled = roll(flat, shifts, 0)
+            return view(rolled, list(a.get_size()))
+        if len_shifts != len_dims:
+            raise RuntimeError(
+                f"shifts and dimensions must align. shifts: {len_shifts}, dims: {len_dims}"
+            )
+        assert len_dims > 1
+        tail_shifts = shifts[1:]
+        tail_dims = dims[1:]
+        first_dim_rolled = roll(a, shifts[0], dims[0])
+        return roll(first_dim_rolled, tail_shifts, tail_dims)
+
+    (dim,) = dims
+    size = V.graph.sizevars.guard_static_shape(a.get_size()[dim])
+    start = (size - shifts[0]) % size
```

Comment:
Will this be a `ZeroDivisionError` if size is 0? 

The aten op does this:

```
if (self.numel() == 0) {
    return self.clone(at::MemoryFormat::Preserve);
  }
```
Does something upstream protect us from size being 0?

Reply:
I don't believe size 0 is supported properly yet in inductor. I can add this though.

Reply:
I guess we could have a shortcut for zero-element tensor to return itself.